**AKBANK VERİ ANALİZİ BOOTCAMP PROJESİ**

**1.VERİSETİNİN KAGGLE PLATFORMUNDAN AKTARILMASI**

In [122]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = "/content/gdrive/My Drive/Kaggle"

In [123]:
%cd /content/gdrive/My Drive/Kaggle

/content/gdrive/My Drive/Kaggle


In [124]:
!kaggle datasets download -d raminhuseyn/airline-customer-satisfaction

Dataset URL: https://www.kaggle.com/datasets/raminhuseyn/airline-customer-satisfaction
License(s): Apache 2.0
airline-customer-satisfaction.zip: Skipping, found more recently modified local copy (use --force to force download)


In [99]:
!unzip \*.zip  && rm *.zip

Archive:  airline-customer-satisfaction.zip
replace Airline_customer_satisfaction.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

In [100]:
import pandas as pd

dataset = pd.read_csv("Airline_customer_satisfaction.csv")
dataset.head()

satisfaction   Customer Type  Age   Type of Travel     Class  \
0    satisfied  Loyal Customer   65  Personal Travel       Eco   
1    satisfied  Loyal Customer   47  Personal Travel  Business   
2    satisfied  Loyal Customer   15  Personal Travel       Eco   
3    satisfied  Loyal Customer   60  Personal Travel       Eco   
4    satisfied  Loyal Customer   70  Personal Travel       Eco   

   Flight Distance  Seat comfort  Departure/Arrival time convenient  \
0              265             0                                  0   
1             2464             0                                  0   
2             2138             0                                  0   
3              623             0                                  0   
4              354             0                                  0   

   Food and drink  Gate location  ...  Online support  Ease of Online booking  \
0               0              2  ...               2                       3   
1               0              3  ...               2                       3   
2               0              3  ...               2                       2   
3               0              3  ...               3                       1   
4               0              3  ...               4                       2   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 3                 0                 3                5   
1                 4                 4                 4                2   
2                 3                 3                 4                4   
3                 1                 0                 1                4   
4                 2                 0                 2                4   

   Cleanliness  Online boarding  Departure Delay in Minutes  \
0            3                2                           0   
1            3                2                         310   
2            4                2                           0   
3            1                3                           0   
4            2                5                           0   

   Arrival Delay in Minutes  
0                       0.0  
1                     305.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 22 columns]

Yaralanılan kaynaklar: https://burcukoculu.medium.com/google-colabde-kaggle-veri-setini-nas%C4%B1l-%C3%A7ekip-%C3%BCzerinde-%C3%A7al%C4%B1%C5%9Fabiliriz-fd983434fcdd

**2.VERİ SETİ**

Proje kapsamında kullanılan veri seti, adı açıklanmayan bir havayolu şirketindeki müşteri memnuniyeti düzeylerine ilişkin aşağıdaki değişkenlere ait bilgiler sağlıyor.
1. Satisfaction (Memnuniyet): Müşterinin memnuniyet düzeyini gösterir.
2. Customer Type (Müşteri tipi):  'Sadık Müşteri' veya 'Sadakatsiz Müşteri'
3. Age (Yaş): Müşterinin yaşı.
4. Type of Travel (Seyahat türü): Seyahatin amacı: 'İş seyahati' veya 'Kişisel Seyahat'.
5. Class (Seyahat sınıfı): 'Business', 'Eco' veya 'Eco Plus'.
6. Flight Distance (Uçuş mesafesi): Kilometre cinsinden uçuşun mesafesi
7. Seat comfort (Koltuk konforu): Uçuş sırasında sağlanan koltuk konforunun derecesi (1'den 5'e).
8. Departure/Arrival time convenient (Kalkış/Varış saati uygunluğu): Kalkış/varış saatinin uygunluğunun değerlendirilmesi (1'den 5'e kadar).
9. Food and drink (Yiyecek ve içecek): Uçuş sırasında sunulan yiyecek ve içecek kalitesine ilişkin derecelendirme (1'den 5'e kadar).
10. Gate location (Kapı konumu): Kapı konumu uygunluğunun derecelendirmesi (1'den 5'e).
11. Inflight wifi service (Uçak içi Wi-Fi hizmeti): Uçak içi Wi-Fi hizmeti memnuniyet derecesi (1'den 5'e).
12. Inflight entertainment (Uçak içi eğlence): Uçak içi eğlenceden memnuniyet derecesi (1'den 5'e).
13. Online support (Çevrimiçi destek): Çevrimiçi müşteri desteği memnuniyetinin derecelendirmesi (1'den 5'e).
14. Ease of Online booking (Online rezervasyon kolaylığı): Çevrimiçi rezervasyon memnuniyeti kolaylığının derecelendirmesi (1'den 5'e kadar).
15. On-board service (Araç içi servis): Araç içi hizmet memnuniyeti derecelendirmesi (1'den 5'e kadar).
16. Leg room service (Bacak mesafesi servisi): Bacak mesafesi hizmetinden memnuniyet derecesi (1'den 5'e).
17. Baggage handling (Bagaj taşıma): Bagaj taşıma memnuniyet derecesi (1'den 5'e kadar).
18. Checkin service (Check-in hizmeti):Check-in hizmetinden memnuniyet derecesi (1'den 5'e).
19. Cleanliness (Temizlik):	Temizlik memnuniyeti derecesi (1 ila 5).
20. Online boarding (Çevrimiçi biniş): Çevrimiçi biniş memnuniyeti derecelendirmesi (1'den 5'e kadar).
21. Departure Delay in Minutes (Dakika Olarak Kalkış Gecikmesi): Dakika cinsinden toplam kalkış gecikmesi.
22. Arrival Delay in Minutes (Dakika Olarak Varış Gecikmesi):	Dakika cinsinden toplam varış gecikmesi.




Veri seti, 22 değişken (sütun) ve 129880 satırdan oluşmaktadır.

Yararlanılan kaynaklar: https://www.kaggle.com/datasets/raminhuseyn/airline-customer-satisfaction/data

**3. KEŞİFSEL VERİ ANALİZİ (EDA)**

In [101]:
dataset.describe()

Age  Flight Distance   Seat comfort  \
count  129880.000000    129880.000000  129880.000000   
mean       39.427957      1981.409055       2.838597   
std        15.119360      1027.115606       1.392983   
min         7.000000        50.000000       0.000000   
25%        27.000000      1359.000000       2.000000   
50%        40.000000      1925.000000       3.000000   
75%        51.000000      2544.000000       4.000000   
max        85.000000      6951.000000       5.000000   

       Departure/Arrival time convenient  Food and drink  Gate location  \
count                      129880.000000   129880.000000  129880.000000   
mean                            2.990645        2.851994       2.990422   
std                             1.527224        1.443729       1.305970   
min                             0.000000        0.000000       0.000000   
25%                             2.000000        2.000000       2.000000   
50%                             3.000000        3.000000       3.000000   
75%                             4.000000        4.000000       4.000000   
max                             5.000000        5.000000       5.000000   

       Inflight wifi service  Inflight entertainment  Online support  \
count          129880.000000           129880.000000   129880.000000   
mean                3.249130                3.383477        3.519703   
std                 1.318818                1.346059        1.306511   
min                 0.000000                0.000000        0.000000   
25%                 2.000000                2.000000        3.000000   
50%                 3.000000                4.000000        4.000000   
75%                 4.000000                4.000000        5.000000   
max                 5.000000                5.000000        5.000000   

       Ease of Online booking  On-board service  Leg room service  \
count           129880.000000     129880.000000     129880.000000   
mean                 3.472105          3.465075          3.485902   
std                  1.305560          1.270836          1.292226   
min                  0.000000          0.000000          0.000000   
25%                  2.000000          3.000000          2.000000   
50%                  4.000000          4.000000          4.000000   
75%                  5.000000          4.000000          5.000000   
max                  5.000000          5.000000          5.000000   

       Baggage handling  Checkin service    Cleanliness  Online boarding  \
count     129880.000000    129880.000000  129880.000000    129880.000000   
mean           3.695673         3.340807       3.705759         3.352587   
std            1.156483         1.260582       1.151774         1.298715   
min            1.000000         0.000000       0.000000         0.000000   
25%            3.000000         3.000000       3.000000         2.000000   
50%            4.000000         3.000000       4.000000         4.000000   
75%            5.000000         4.000000       5.000000         4.000000   
max            5.000000         5.000000       5.000000         5.000000   

       Departure Delay in Minutes  Arrival Delay in Minutes  
count               129880.000000             129487.000000  
mean                    14.713713                 15.091129  
std                     38.071126                 38.465650  
min                      0.000000                  0.000000  
25%                      0.000000                  0.000000  
50%                      0.000000                  0.000000  
75%                     12.000000                 13.000000  
max                   1592.000000               1584.000000

Havayoluna ait memnuniyet anketine katılanların 7 ile 85 yaş aralığında oldukları görülmektedir.

In [102]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129880 entries, 0 to 129879
Data columns (total 22 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   satisfaction                       129880 non-null  object 
 1   Customer Type                      129880 non-null  object 
 2   Age                                129880 non-null  int64  
 3   Type of Travel                     129880 non-null  object 
 4   Class                              129880 non-null  object 
 5   Flight Distance                    129880 non-null  int64  
 6   Seat comfort                       129880 non-null  int64  
 7   Departure/Arrival time convenient  129880 non-null  int64  
 8   Food and drink                     129880 non-null  int64  
 9   Gate location                      129880 non-null  int64  
 10  Inflight wifi service              129880 non-null  int64  
 11  Inflight entertainment             1298

In [103]:
def column_detection(dataframe: pd.DataFrame, cat_th: int = 10, car_th: int = 20):
    cat_cols = [col for col in dataframe.columns if dataframe[col].dtypes == "O"]
    num_but_cat = [col for col in dataframe.columns if dataframe[col].nunique() < cat_th and
                       dataframe[col].dtypes != "O"]
    cat_but_car = [col for col in dataframe.columns if dataframe[col].nunique() > car_th and
                    dataframe[col].dtypes == "O"]
    cat_cols = cat_cols + num_but_cat
    cat_cols = [col for col in cat_cols if col not in cat_but_car]

    # num_cols
    num_cols = [col for col in dataframe.columns if dataframe[col].dtypes != "O"]
    num_cols = [col for col in num_cols if col not in num_but_cat]

    print("\n##################### Categoric #####################")
    print(cat_cols)
    print("\n##################### Numeric #####################")
    print(num_cols)
    print("\n##################### Categoric But Cardinal #####################")
    print(cat_but_car)
    print("\n##################### Numeric But Categoric #####################")
    print(num_but_cat)
    print("\n##################### Columns Overview #####################")
    print(f"Observations: {dataframe.shape[0]}")
    print(f"Variables: {dataframe.shape[1]}")
    print(f'cat_cols: {len(cat_cols)}')
    print(f'num_cols: {len(num_cols)}')
    print(f'cat_but_car: {len(cat_but_car)}')
    print(f'num_but_cat: {len(num_but_cat)}')
    return cat_cols, num_cols, cat_but_car, num_but_cat

In [104]:
cat_cols, num_cols, cat_but_car, num_but_cat = column_detection(dataset)


##################### Categoric #####################
['satisfaction', 'Customer Type', 'Type of Travel', 'Class', 'Seat comfort', 'Departure/Arrival time convenient', 'Food and drink', 'Gate location', 'Inflight wifi service', 'Inflight entertainment', 'Online support', 'Ease of Online booking', 'On-board service', 'Leg room service', 'Baggage handling', 'Checkin service', 'Cleanliness', 'Online boarding']

##################### Numeric #####################
['Age', 'Flight Distance', 'Departure Delay in Minutes', 'Arrival Delay in Minutes']

##################### Categoric But Cardinal #####################
[]

##################### Numeric But Categoric #####################
['Seat comfort', 'Departure/Arrival time convenient', 'Food and drink', 'Gate location', 'Inflight wifi service', 'Inflight entertainment', 'Online support', 'Ease of Online booking', 'On-board service', 'Leg room service', 'Baggage handling', 'Checkin service', 'Cleanliness', 'Online boarding']

##############

* Veri setinde;
 - 18 tane kategorik,
 - 4 tane sayısal (nicel)

 değişken olduğu görülmektedir. Bu değişkenlerin 14 tanesi sayısal olarak girilmiş kategorik değişken olarak görülmektedir. Veri setinde "Seat comfort" ile "Online boarding" değişenleri arasındaki değişkenlerin veri tipi sayısal olarak görülmektedir. Bu değişkenlerin veri tipleri kategorik olarak değiştirilmelidir.

* Veri setindeki bütün değişkenler 129880 satırdan oluşurken son değişken olan dakika olarak belirtilen kalkış gecikmesi değişkeninin 129487 satırdan oluştuğu görülmektedir. Belirtilen değişkende eksik değerlerin olup olmadığı incelenmelidir.

In [105]:
for column in dataset.columns:
    parametres = dataset[column].value_counts()
    print(parametres, end='\n')
    print(sum(parametres), end='\n\n')

satisfaction
satisfied       71087
dissatisfied    58793
Name: count, dtype: int64
129880

Customer Type
Loyal Customer       106100
disloyal Customer     23780
Name: count, dtype: int64
129880

Age
39    3692
25    3511
40    3209
44    3104
41    3089
      ... 
74      61
76      60
79      52
78      44
85      25
Name: count, Length: 75, dtype: int64
129880

Type of Travel
Business travel    89693
Personal Travel    40187
Name: count, dtype: int64
129880

Class
Business    62160
Eco         58309
Eco Plus     9411
Name: count, dtype: int64
129880

Flight Distance
1963    92
1812    88
1639    87
1981    86
1789    86
        ..
4222     1
5049     1
5378     1
5613     1
4260     1
Name: count, Length: 5398, dtype: int64
129880

Seat comfort
3    29183
2    28726
4    28398
1    20949
5    17827
0     4797
Name: count, dtype: int64
129880

Departure/Arrival time convenient
4    29593
5    26817
3    23184
2    22794
1    20828
0     6664
Name: count, dtype: int64
129880

Food and 

* "satisfaction" ve "Customer Type" değişkenin ve kategorilerinin isimleri düzenlenebilir. (Büyük/Küçük harf uyumu sağlanabilir.)
* "Arrival Delay in Minutes" ve "Departure Delay in Minutes" değişkenleri aynı veri türüne sahip olmalıdır.

In [108]:
print(dataset.isnull().sum())

satisfaction                           0
Customer Type                          0
Age                                    0
Type of Travel                         0
Class                                  0
Flight Distance                        0
Seat comfort                           0
Departure/Arrival time convenient      0
Food and drink                         0
Gate location                          0
Inflight wifi service                  0
Inflight entertainment                 0
Online support                         0
Ease of Online booking                 0
On-board service                       0
Leg room service                       0
Baggage handling                       0
Checkin service                        0
Cleanliness                            0
Online boarding                        0
Departure Delay in Minutes             0
Arrival Delay in Minutes             393
dtype: int64


Veri setinde "Arrival Delay in Minutes (Dakika Olarak Varış Gecikmesi)" değişkeninde 393 adet eksik veri olduğu ve diğer değişkenler de eksik verinin olmadığı görülmektedir.

Yararlanılan kaynaklar: https://github.com/Enes-CE/Netflix-Data-Analys/blob/main/Netflix_DataAnalys.ipynb

https://gokerguner.medium.com/machine-learning-1-7d4581caa291

**4.VERİ ÖN İŞLEME**

Yapılacaklar:

1. "Seat comfort" ile "Online boarding" değişenleri arasındaki değişkenlerin  veri tipleri kategorik olarak değiştirilecek.
2. "Arrival Delay in Minutes" değişkeninde bulunan eksik değerler doldurulacak.
3. Değişkenlerin isimlerinin baş harfleri büyük harfle başlayacak şekilde düzenlenecek.
4. "Departure Delay in Minutes" değişkeninin veri tipi "float" olarak değiştirilecek.


In [111]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129880 entries, 0 to 129879
Data columns (total 22 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   satisfaction                       129880 non-null  object 
 1   Customer Type                      129880 non-null  object 
 2   Age                                129880 non-null  int64  
 3   Type of Travel                     129880 non-null  object 
 4   Class                              129880 non-null  object 
 5   Flight Distance                    129880 non-null  int64  
 6   Seat comfort                       129880 non-null  int64  
 7   Departure/Arrival time convenient  129880 non-null  int64  
 8   Food and drink                     129880 non-null  int64  
 9   Gate location                      129880 non-null  int64  
 10  Inflight wifi service              129880 non-null  int64  
 11  Inflight entertainment             1298

In [114]:
for i in range(5,19):
  dataset.iloc[:, i+1] = dataset.iloc[:, i+1].astype('category')

dataset['Departure Delay in Minutes'] = dataset['Departure Delay in Minutes'].astype('float')

dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129880 entries, 0 to 129879
Data columns (total 22 columns):
 #   Column                             Non-Null Count   Dtype   
---  ------                             --------------   -----   
 0   satisfaction                       129880 non-null  object  
 1   Customer Type                      129880 non-null  object  
 2   Age                                129880 non-null  int64   
 3   Type of Travel                     129880 non-null  object  
 4   Class                              129880 non-null  object  
 5   Flight Distance                    129880 non-null  int64   
 6   Seat comfort                       129880 non-null  category
 7   Departure/Arrival time convenient  129880 non-null  category
 8   Food and drink                     129880 non-null  category
 9   Gate location                      129880 non-null  category
 10  Inflight wifi service              129880 non-null  category
 11  Inflight entertainment    

Veri setindeki değişkenlerin içerdikleri verilere göre veri türleri düzenlenmiştir.

In [115]:
dataset.columns

Index(['satisfaction', 'Customer Type', 'Age', 'Type of Travel', 'Class',
       'Flight Distance', 'Seat comfort', 'Departure/Arrival time convenient',
       'Food and drink', 'Gate location', 'Inflight wifi service',
       'Inflight entertainment', 'Online support', 'Ease of Online booking',
       'On-board service', 'Leg room service', 'Baggage handling',
       'Checkin service', 'Cleanliness', 'Online boarding',
       'Departure Delay in Minutes', 'Arrival Delay in Minutes'],
      dtype='object')

In [116]:
dataset.columns = dataset.columns.str.title()
dataset.columns

Index(['Satisfaction', 'Customer Type', 'Age', 'Type Of Travel', 'Class',
       'Flight Distance', 'Seat Comfort', 'Departure/Arrival Time Convenient',
       'Food And Drink', 'Gate Location', 'Inflight Wifi Service',
       'Inflight Entertainment', 'Online Support', 'Ease Of Online Booking',
       'On-Board Service', 'Leg Room Service', 'Baggage Handling',
       'Checkin Service', 'Cleanliness', 'Online Boarding',
       'Departure Delay In Minutes', 'Arrival Delay In Minutes'],
      dtype='object')

Değişken isimlerinde büyük/küçük harf uyumu sağlanmıştır.

In [117]:
#Eksik değer tablosu
def eksik_deger_tablosu(df):
    eksik_deger = df.isnull().sum()
    eksik_deger_yuzde = 100 * df.isnull().sum()/len(df)
    eksik_deger_tablo = pd.concat([eksik_deger, eksik_deger_yuzde], axis=1)
    eksik_deger_tablo_son = eksik_deger_tablo.rename(
    columns = {0 : 'Eksik Değerler', 1 : '% Değeri'})
    return eksik_deger_tablo_son

In [118]:
 eksik_deger_tablosu(dataset)

Eksik Değerler  % Değeri
Satisfaction                                    0  0.000000
Customer Type                                   0  0.000000
Age                                             0  0.000000
Type Of Travel                                  0  0.000000
Class                                           0  0.000000
Flight Distance                                 0  0.000000
Seat Comfort                                    0  0.000000
Departure/Arrival Time Convenient               0  0.000000
Food And Drink                                  0  0.000000
Gate Location                                   0  0.000000
Inflight Wifi Service                           0  0.000000
Inflight Entertainment                          0  0.000000
Online Support                                  0  0.000000
Ease Of Online Booking                          0  0.000000
On-Board Service                                0  0.000000
Leg Room Service                                0  0.000000
Baggage Handling                                0  0.000000
Checkin Service                                 0  0.000000
Cleanliness                                     0  0.000000
Online Boarding                                 0  0.000000
Departure Delay In Minutes                      0  0.000000
Arrival Delay In Minutes                      393  0.302587

Veri setinde bulunan "Arrival Delay In Minutes" değişkeninde 393 tane eksik değerin ve değişkenin %30'unun bu eksik değerlerden oluştuğu görülmektedir. Bu değişken sürekli değerler aldığı için eksik değerlerin doldurulması ortalama, medyan, mod ya da lineer regresyon ile yapılmalıdır.

In [119]:
dataset['Arrival Delay In Minutes'] = dataset['Arrival Delay In Minutes'].interpolate(method="linear")
eksik_deger_tablosu(dataset)

Eksik Değerler  % Değeri
Satisfaction                                    0       0.0
Customer Type                                   0       0.0
Age                                             0       0.0
Type Of Travel                                  0       0.0
Class                                           0       0.0
Flight Distance                                 0       0.0
Seat Comfort                                    0       0.0
Departure/Arrival Time Convenient               0       0.0
Food And Drink                                  0       0.0
Gate Location                                   0       0.0
Inflight Wifi Service                           0       0.0
Inflight Entertainment                          0       0.0
Online Support                                  0       0.0
Ease Of Online Booking                          0       0.0
On-Board Service                                0       0.0
Leg Room Service                                0       0.0
Baggage Handling                                0       0.0
Checkin Service                                 0       0.0
Cleanliness                                     0       0.0
Online Boarding                                 0       0.0
Departure Delay In Minutes                      0       0.0
Arrival Delay In Minutes                        0       0.0

In [120]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129880 entries, 0 to 129879
Data columns (total 22 columns):
 #   Column                             Non-Null Count   Dtype   
---  ------                             --------------   -----   
 0   Satisfaction                       129880 non-null  object  
 1   Customer Type                      129880 non-null  object  
 2   Age                                129880 non-null  int64   
 3   Type Of Travel                     129880 non-null  object  
 4   Class                              129880 non-null  object  
 5   Flight Distance                    129880 non-null  int64   
 6   Seat Comfort                       129880 non-null  category
 7   Departure/Arrival Time Convenient  129880 non-null  category
 8   Food And Drink                     129880 non-null  category
 9   Gate Location                      129880 non-null  category
 10  Inflight Wifi Service              129880 non-null  category
 11  Inflight Entertainment    

Veri setinde bulunan "Arrival Delay In Minutes" değişkenindeki kayıp değerler lineer bir interpolasyon yöntemi kullanılarak doldurulmuştur. Böylelikle artık veri setinde herhangi bir değişkene ilişkin kayıp değerler bulunmamaktadır.

Yararlanılan kaynaklar: https://yigitsener.medium.com/veri-biliminde-eksik-kay%C4%B1p-verilere-yakla%C5%9F%C4%B1m-stratejileri-ve-python-pandas-uygulamas%C4%B1-501fbf643795

**5.İSTATİSTİKSEL ANALİZLER**

In [ ]:
dataset.describe()

İstatistiksel analizler incelendiğinde havayolu şirketinin yaptığı ankete katılan insanların;

* en fazla 85, en az 7 yaşında olduklarını ve ortalama olarak 39 yaşındaki insanların ankete katılım sağladıklarını,
* en fazla 6951 km, en az 50 km uçuş yaptıklarını ve ankete katılım sağlayan insanların ortalama olarak 1981 km uçuş yaptıklarını,
* bilet aldıkları uçakların ya hiç gecikme olmadan ya da en fazla 1592 dakika gecikmeli olarak kalkış yaptığını,
* bilet aldıkları uçakların ya hiç gecikme olmadan ya da en fazla 1584 dakika gecikmeli olarak gidecekleri yere varış yaptığını

söyleyebiliriz.

**NOT:** İstatistiksel analizler sadece nicel yani sayısal verilere uygulanmıştır. Kategorik verilerin ortalaması hesaplanamaz, çünkü kategorik verilerde sayısal bir değer bulunmaz. Bunun yerine, kategorik verilerin frekansı (yani her kategorideki gözlem sayısı) hesaplanabilir. Kategorilerin her birinin frekansı, her bir kategorinin kaç kez gözlemlendiğini gösterir. Frekans analizi, kategorik verilerin dağılımını ve eğilimlerini anlamak için kullanılır.

**6.VERİ GÖRSELLEŞTİRME**

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

pastel_palette = sns.color_palette("pastel")

# Sadece kategorik değişkenler için pasta grafikleri çizelim:
for variable in dataset.select_dtypes(include=['object', 'category']):
    plt.figure(figsize=(6, 6))
    dataset[variable].value_counts().plot.pie(autopct='%1.1f%%', colors=pastel_palette)
    plt.title(variable)
    plt.ylabel('')  # Y ekseninin etiketini kaldırır
    plt.show()

Kategorik değişkenlerine ait pasta grafikleri incelendiğinde;

* Müşterilerin %54.7'sinin havayolu firmasından memnun olduklarını,
* Müşterilerin %81.7'sinin sadık müşteri olduklarını,
* Müşterilerin %69.1'inin iş seyahati için havayolu firmasını kullandıklarını,
* Müşterilerin %47.9'unun business ve %44.9'unun ekonomi sınıfı olduklarını,
* Müşterilerin %22.5'inin uçağın koltuk konforuna 5 üzerinden 3 puan verdiklerini yani uçağın koltuk konforunun orta derece iyi olduğunu,
* Müşterilerin %22.8'inin uçağın kalkış ve varış saatlerinin uygunluğuna 5 üzerinden 4 puan verdiklerini yani uçağın kalkış ve varış saatinin genellikle uygun olduğunu,
* Müşterilerin %21.7'sinin uçuş esnasında verilern yiyecek ve içeceklerin kalitesine 5 üzerinden 3 puan verdiklerini yani uçuş esnasında verilen yiyecek ve içeceklerin orta derecede kaliteli olduğunu,
* Müşterilerin %25.8'inin uçağın kapı konumunun uygunluğuna 5 üzerinden 3 puan verdiklerini yani uçağın kapı konumunun orta derecede uygun olduğunu,
* Müşterilerin %24.3'ünün uçağın içindeki wifi servisine 5 üzerinden 4 puan verdiklerini yani uçağın içindeki wifi servisinden genellikle memnun olduklarını,
* Müşterilerin %32.2'sinin uçağın içindeki eğlenceye 5 üzerinden 4 puan verdiklerini yani uçağın içindeki eğlenceden genellikle memnun olduklarını,
* Müşterilerin %32'sinin uçağın çevrimiçi desteğine 5 üzerinden 4 puan verdiklerini yani uçağın çevrimiçi desteğinden genellikle memnun olduklarını,
* Müşterilerin %30.7'sinin uçağın çevrimiçi rezervasyon kolaylığına 5 üzerinden 4 puan verdiklerini yani uçağın çevrimiçi rezervasyon kolaylığından genellikle memnun olduklarını,
* Müşterilerin %31.3'ünün uçağın içindeki servis hizmetine 5 üzerinden 4 puan verdiklerini yani uçağın içindeki servis hizmetinden genellikle memnun olduklarını,
* Müşterilerin %30.6'ünün uçağın içindeki bacak mesafesinin uygunluğuna 5 üzerinden 4 puan verdiklerini yani uçağın içindeki bacak mesafesinden genellikle memnun olduklarını,
* Müşterilerin %37.1'inin uçağın bagaj taşıma memuniyetine 5 üzerinden 4 puan verdiklerini yani uçağın bagaj taşımasından genellikle memnun olduklarını,
* Müşterilerin %28.1'inin uçağın check-in hizmetine 5 üzerinden 4 puan verdiklerini yani uçağın check-in hizmetinden genellikle memnun olduklarını,
* Müşterilerin %37.6'sının uçağın temizliğine 5 üzerinden 4 puan verdiklerini yani uçağın temizliğinden genellikle memnun olduklarını,
* Müşterilerin %27.1'inin uçağın çevrim içi biniş hizmetine 5 üzerinden 4 puan verdiklerini yani uçağın çevrim içi biniş hizmetinden genellikle memnun olduklarını

söyleyebiliriz.

In [ ]:
# Her bir nicel değişken için bir döngü oluşturalım:
for column in dataset.select_dtypes(include=['int', 'float']):
    # Değişkenin dağılımını alıp histogram olarak çizelim
    plt.figure(figsize=(8, 6))
    plt.hist(dataset[column], color='skyblue', bins=20)  # bins parametresiyle bar sayısını ayarlayabilirsiniz
    plt.title(f'Distribution of {column}')
    plt.xlabel(column)
    plt.ylabel('Frequency')
    plt.show()

Nicel değişkenlerin dağılımlarına ait histogram grafikleri incelendiğinde;

* Müşterilerin yaş dağılımlarının normal değılıma sahip olduğunu yani yaşın geniş bir aralıkta ve neredeyse simetrik bir şekilde dağıldığını söyleyebiliriz. Bu durum, yaş dağılımının istatistiksel olarak normal bir yapıya sahip olduğunu ve veri setinde yaşlı ve genç müşterilerin dengeli bir şekilde temsil edildiğini gösterir.
* Müşterilerin uçuş mesafelerinin dağılımının sağa çarpık olduğunu yani uçuş mesafesinin büyük çoğunluğunun kısa mesafelerde olduğunu ancak bazı müşterilerin çok daha uzun mesafelere seyahat ettiğini söyleyebiliriz. Sağa çarpık bir dağılım, ortalamadan daha büyük değerlerin daha fazla olduğunu ve dağılımın kuyruğunun sağa doğru uzadığını ifade eder. Bu durum, seyahat mesafesinin genellikle kısa olduğu ancak nadiren uzun seyahatlerin yapıldığını gösterir.
* Uçağın dakika olarak kalkış ve varış gecikmesinin dağılımlarının da sağa çarpık olduğunu yani çoğu durumda uçuşların planlandığı zamandan önce kalkış yaptığını, ancak nadiren planlandığı zamandan daha geç kalkışlar olduğunu ve benzer şekilde, çoğu durumda uçuşların planlandığı zamanda varış yaptığını ancak bazı durumlarda varışın planlanandan daha geç olduğunu söyeleyebiliriz.

**7.DEĞİŞKENLER ARASI İLİŞKİ ANALİZİ**

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Veri setindeki kategorik değişkenlerin adlarını alalım
categorical_columns = dataset.select_dtypes(include=['object','category']).columns

# LabelEncoder nesnesini oluşturalım
label_encoder = LabelEncoder()

# Her bir kategorik değişkeni label encoding ile dönüştürelim
for column in categorical_columns:
    dataset[column] = label_encoder.fit_transform(dataset[column])

dataset.describe()

In [ ]:
# Belirli bir eşik değeri belirleyelim
threshold = 0.5

# Korelasyon matrisindeki sadece eşik değerinden büyük korelasyon katsayılarını alalım
# Korelasyon matrisini oluşturalım
corr_matrix = dataset.corr().abs()
high_corr = corr_matrix[corr_matrix > threshold]

# Isı haritasını çizelim
plt.figure(figsize=(12, 8))
sns.heatmap(high_corr, annot=True, cmap='GnBu', linewidths=0.2, vmin=0, vmax=1)
plt.xlabel('Özellikler')
plt.ylabel('Özellikler')
plt.title('Korelasyon Matrisi (Eşik Değerinden Büyük Korelasyonlar)')
plt.show()


Kategorik olan veriler label encoder ile sayısal veriye dönüştürülmüştür. Değişkenler arasındaki ilişki 0.5 eşik değeri olarak belirlenip eşik değerinden büyük ilişkisi olan değişkenler ekrana yansıtılmıştır. Korelasyon grafiği incelendiğinde en yüksek ilişkiye uçağın dakika cinsinden varış ve kalkış gecikmesi olduğunu ve aralarında %96'lık bir ilişki olduğu görülmektedir.

**8.FEATURE SELECTION**

LASSO (En Az Mutlak Büzülme ve Seçim Operatörü), aşırı uyumu önlemek için hem özellik seçimini hem de düzenlemeyi gerçekleştiren doğrusal regresyon yönteminin bir modifikasyonudur. LASSO, daha az önemli özelliklerin regresyon katsayılarını sıfıra indirerek bunları modelden etkili bir şekilde çıkarır. Geriye kalan sıfır olmayan katsayılar önemli özellikleri göstermektedir.

In [ ]:
from sklearn.linear_model import LassoCV

# Hedef değişkeni tanımlayın
target = dataset['Satisfaction']

# Veri setinden hedef değişkeni çıkarın
X = dataset.drop(columns=['Satisfaction'])

# LASSO regresyon modelini oluşturalım
lasso = LassoCV(cv=5, random_state=0)
lasso.fit(X, target)

# Katsayıları çizelim
plt.figure(figsize=(10, 6))
plt.plot(range(len(X.columns)), lasso.coef_, marker='o', markersize=8, linestyle='None')
plt.axhline(y=0, color='gray', linestyle='--', linewidth=2)
plt.xticks(range(len(X.columns)), X.columns, rotation=90)
plt.ylabel('Katsayılar')
plt.xlabel('Özellikler')
plt.title('LASSO Regresyonunu Kullanarak Özelliğin Önemi')
plt.show()

Elde edilen grafik incelendiğinde;
* Uçak içi eğlence değişkeninin müşteri memnuniyeti üzerinde en fazla etkiye sahip olduğu,
* Kalkış/Varış saati uygunluğu değişkeninin ise müşteri memnuniyeti üzerinde en az etkiye sahip olduğu

gözlemlenmiştir.

Lasso regresyon ile müşteri memnuniyeti üzerinde etkisi olmayan değişkenlerin ise katsayısı "0" olarak atanmıştır. Bu değişkenler;
* Yaş
* Seyahat türü
* Uçuş mesafesi
* Yiyecek ve içecek kalitesi memnuniyeti
* Kapı konumu memnuniyeti
* Uçak içi wifi hizmeti memnuniyeti
* Bağaj taşıma memnuniyeti
* Temizlik memnuniyeti
* Çevrimiçi biniş memnuniyeti
* Dakika olarak kalkış gecikmesi
* Dakika olarak varış gecikmesi

Yararlanılan Kaynaklar: https://medium.com/@daython3/mastering-the-art-of-feature-selection-python-techniques-for-visualizing-feature-importance-cacf406e6b71

**9.FEATURE ENGINEERING**

In [ ]:
# Lasso regresyon ile katsayısı 0 olan değişkenler çıkarılarak yeni bir DataFrame oluşturulur
selected_dataset = dataset.drop(columns=['Age', 'Type Of Travel', 'Flight Distance', 'Food And Drink', 'Gate Location', 'Inflight Wifi Service', 'Baggage Handling', 'Cleanliness', 'Online Boarding', 'Departure Delay In Minutes', 'Arrival Delay In Minutes'])

selected_dataset.info()

Lasso regresyon kullanılarak belirlenen önemsiz değişkenler yeni oluşturulan veri setine dahil edilmemiştir.

Yararlanılan Kaynaklar: https://medium.com/@daython3/mastering-the-art-of-feature-selection-python-techniques-for-visualizing-feature-importance-cacf406e6b71